akka-actor/src/main/scala/akka/routing/DefaultMetricsBasedResizer.scala

```diff
@@ -0,0 +1,221 @@
+import scala.util.Random
+
+import com.typesafe.config.Config
+
+import akka.actor._
+
+import DefaultMetricsBasedResizer._
+
+trait MetricsBasedResizer extends Resizer {
+  /**
+   * Report the messageCount as well as current routees so that the
+   * it can collect metrics.
+   * Caution: this method is not thread safe.
+   *
+   * @param currentRoutees
+   * @param messageCounter
+   */
+  def reportMessageCount(currentRoutees: immutable.IndexedSeq[Routee], messageCounter: Long): Unit
+}
+
+case object DefaultMetricsBasedResizer {
+  private[routing]type PoolSize = Int
+
+  private[routing] case class UnderUtilizationStreak(start: LocalDateTime, highestUtilization: Int)
+
+  private[routing] case class ResizeRecord(
+    underutilizationStreak: Option[UnderUtilizationStreak] = None,
+    messageCount: Long = 0,
+    totalQueueLength: Int = 0,
+    checkTime: Option[LocalDateTime] = None)
+
+  private[routing]type PerformanceLog = Map[PoolSize, Duration]
+
+  def apply(resizerCfg: Config): DefaultMetricsBasedResizer =
+    DefaultMetricsBasedResizer(
+      lowerBound = resizerCfg.getInt("lower-bound"),
+      upperBound = resizerCfg.getInt("upper-bound"),
+      chanceOfScalingDownWhenFull = resizerCfg.getDouble("chance-of-ramping-down-when-full"),
+      actionInterval = resizerCfg.getDuration("action-interval"),
+      downsizeAfterUnderutilizedFor = resizerCfg.getDuration("downsize-after-underutilized-for"),
+      exploreStepSize = resizerCfg.getDouble("explore-step-size"),
+      explorationProbability = resizerCfg.getDouble("chance-of-exploration"),
+      weightOfLatestMetric = resizerCfg.getDouble("weight-of-latest-metric"),
+      downsizeRatio = resizerCfg.getDouble("downsize-ratio"))
+
+}
+
+/**
+ * Implementation of [[Resizer]] that adjust the [[Pool]]
+ * based on performance per size.
+ * It keeps a performance log so it's stateful as well as
+ * a larger memory footprint than the [[DefaultResizer]].
+ * For documentation about the the parameters, see the reference.conf -
+ * akka.actor.deployment.default.metrics-based-resizer
+ *
+ * CAUTION: [[resize]] also changes the inner state of the resizer,
+ * do not call it unnecessarily or even concurrently
+ *
+ *
+ */
+@SerialVersionUID(1L)
+case class DefaultMetricsBasedResizer(
+  lowerBound: PoolSize = 1,
+  upperBound: PoolSize = 30,
+  chanceOfScalingDownWhenFull: Double = 0.2,
+  actionInterval: Duration = Duration.ofSeconds(5),
+  numOfAdjacentSizesToConsiderDuringOptimization: Int = 6,
+  exploreStepSize: Double = 0.1,
+  downsizeRatio: Double = 0.8,
+  downsizeAfterUnderutilizedFor: Duration = Duration.ofHours(72),
+  explorationProbability: Double = 0.4,
+  weightOfLatestMetric: Double = 0.5,
+  /**
+ * Leave package accessible for testing purpose
+ */
+  private[routing] var performanceLog: PerformanceLog = Map.empty,
+  /**
+ * Leave package accessible for testing purpose
+ */
+  private[routing] var record: ResizeRecord = ResizeRecord()) extends MetricsBasedResizer {
```

Comment:
Will do. I put them there for convenience when writing unit tests (so that I can create a resizer with these two fields preset in a single statement). Their participation in hashcode/equality didn't worry me.